frontend/src/metabase/browse/utils.ts

```diff
@@ -41,16 +47,18 @@ export type ModelFilterControlsProps = {
  * or false if it is inactive */
 export type ActualModelFilters = Record<string, boolean>;
 
-export const filterModels = (
-  unfilteredModels: ModelResult[],
+export const filterModels = <T extends ModelResult[] | RecentCollectionItem[]>(
+  unfilteredModels: T,
   actualModelFilters: ActualModelFilters,
   availableModelFilters: AvailableModelFilters,
-) => {
+): Array<T[number]> => {
   return _.reduce(
     actualModelFilters,
     (acc, shouldFilterBeActive, filterName) =>
       shouldFilterBeActive
-        ? acc.filter(availableModelFilters[filterName].predicate)
+        ? // https://devblogs.microsoft.com/typescript/announcing-typescript-5-2/#easier-method-usage-for-unions-of-arrays
+          // @ts-expect-error This would be valid without a cast in TS 5.2+
+          (acc.filter(availableModelFilters[filterName].predicate) as T)
```

Comment:
Suggestion (take it or leave it, of course!): Explicitly do the cast instead of using a `@ts-expect-error`: the cast will at least catch it if you do something *entirely* unreasonable, whereas the `@ts-expect-error` will flat-out swallow *any* error (you could type `acc.filter("lol")` and it would still be accepted by the compiler in that case, though presumably not by your code review process).

I *think* this should work:

```suggestion
          (acc.filter(availableModelFilters[filterName].predicate) as Array<T[number]>)
```

(Not sure if it'll pass must on your Prettier settings, though!)

Reply:
sadly, Typescript 4.7 isn't happy with that cast, see [example](https://www.typescriptlang.org/play/?ts=4.7.4#code/MYewdgzgLgBAZgSwDZQKYCcBiYYF4YAUCYADgK5QBcMYZAtgEYYwA+M06xA5gJR4B8MAN4AoGDARxCUAJ4lUIKcXKxcamAHJajDBr6jx49Kihl0OZRRiCArAG4xMAL4jHx0+YmkrgjXBAgGg4uIqCQsCAUGJjIaOh4MAA8ACowqAAeaGAAJhAwAILo6ACGMonaTOiCbIUlZRzc-PxE3lQwyTzUtaUpANoVGAC6griCBjDuZjgtKgB0iCgYBAtx2Dx8xXndZcn99JXDwa5IJjDFCZGrsUu9AIwANDAATIM8DiewDBdRWNfoBL0NMUNI8NAwNK93qdgN8rot-ndQU8IW8gA)

Reply:
Oh, duh, if it were it would accept it in the first place. The problem is actually on the call side for `.filter()`, which does not like the predicate being typed this way. That was improved in 5.2 as well. [Here is an updated playground][1] with safe (!) cast you can use until then, which casts the input array instead of the result.

[1]: https://www.typescriptlang.org/play/?ts=4.7.4#code/MYewdgzgLgBAZgSwDZQKYCcBiYYF4YAUCYADgK5QBcMYZAtgEYYwA+M06xA5gJR4B8MAN4AoGDARxCUAJ4lUIKcXKxcamAHJajDBr6jx49Kihl0OZRRiCArAG4xMAL4jHx0+YmkrgjXBAgGg4uIqCQsCAUGJjIaOh4MAA8ACowqAAeaGAAJhAwAILo6ACGMonaTOiCbIUlZRzc-PxE3lQwyTzUtaUpANoVGAC6griCBjDuZjgELSowxXnd9VCcYFysNPSV-DwAdIgoGAQHcdg8PMGuSCbzCZGnsUe9AIwANDAATIMXItewDHcolhHugCL0NMUNO8NAwNN8HH8YMBAQ9DqCXtCPnCLkA